File: integration/providers/epfl/transcript-fixture.py
#!/usr/bin/env python3
"""EPFL transcript campaign factory for synthetic dc+sd-jwt variants."""
from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path

from epfl_challenge import CIRCUIT_ID, PROFILE, digest_hex, policy_inputs_json
from fixtures import ISSUER_ID, KEY_ID, NOW_DATE, VARIANT_A, VARIANT_B, fixture_bundle

QUERY_ID = "birth_date"
CLIENT_ID = "did:example:epfl-verifier"
RESPONSE_URI = "https://example.test/oid4vp/api/request-object/00000000-0000-4000-8000-000000000001/response-data"
STATE = "epfl-transcript-state-v1"
NONCE = "epfl-transcript-nonce-v1"


def _session_challenge_hex(issuer_pub_x: list[int], issuer_pub_y: list[int]) -> str:
    policy_inputs = policy_inputs_json(
        CIRCUIT_ID,
        "".join(f"{b:02x}" for b in issuer_pub_x),
        "".join(f"{b:02x}" for b in issuer_pub_y),
        NOW_DATE,
    )
    return digest_hex(NONCE, CLIENT_ID, RESPONSE_URI, STATE, QUERY_ID, PROFILE, policy_inputs)


def _variant(variant_id: str, witness) -> dict:
    challenge_hex = _session_challenge_hex(witness.issuer_pub_x, witness.issuer_pub_y)
    bundle = fixture_bundle(variant_id, challenge_hex)
    given = bundle["given"]
    request_context = {
        "issuer_pub_x": witness.issuer_pub_x,
        "issuer_pub_y": witness.issuer_pub_y,
        "now_date": NOW_DATE,
        "challenge_nonce": list(bytes.fromhex(challenge_hex)),
    }
    policy_inputs = policy_inputs_json(
        CIRCUIT_ID,
        "".join(f"{b:02x}" for b in witness.issuer_pub_x),
        "".join(f"{b:02x}" for b in witness.issuer_pub_y),
        NOW_DATE,
    )
    return {
        "id": variant_id,
        "credential": {"format": "dc+sd-jwt", "data": bundle["credential"]},
        "given": given,
        "prepare": {
            "profile": PROFILE,
            "credential": {"format": "epfl-variant", "variant": variant_id},
            "context": {},
        },
        "present": {
            "profile": PROFILE,
            "request_context": request_context,
            "inputs": {},
        },
        "verify": {
            "profile": PROFILE,
            "request_context": request_context,
            "inputs": {},
        },
        "transcript": {
            "nonce": NONCE,
            "client_id": CLIENT_ID,
            "response_uri": RESPONSE_URI,
            "state": STATE,
            "query_id": QUERY_ID,
            "profile": PROFILE,
            "circuit_id": CIRCUIT_ID,
            "policy_inputs": policy_inputs,
            "challenge_nonce_hex": challenge_hex,
        },
        "compact_sd_jwt_sha256": hashlib.sha256(bundle["credential"].encode("ascii")).hexdigest(),
    }


def build_document() -> dict:
    issuer_x = "".join(f"{b:02x}" for b in VARIANT_A.issuer_pub_x)
    issuer_y = "".join(f"{b:02x}" for b in VARIANT_A.issuer_pub_y)
    return {
        "schema": "swiyu.transcript-fixture.v1",
        "equivalence": "not_claimed",
        "mapping": {
            "format": "dc+sd-jwt",
            "profile": PROFILE,
            "query_id": QUERY_ID,
            "challenge_from": "transcript.challenge_nonce_hex",
            "givens": {
                "issuer": "issuer",
                "now_date": "now_date",
                "challenge_nonce": "challenge_nonce",
            },
            "challenge_fields": {
                "nonce": "nonce",
                "clientId": "client_id",
                "responseUri": "response_uri",
                "state": "state",
            },
            "signature_opaque": True,
        },
        "dcql_query": {
            "credentials": [{
                "id": QUERY_ID,
                "format": "dc+sd-jwt",
                "meta": {"vct_values": ["https://example.ch/vct/epfl-d10-test"]},
                "require_cryptographic_holder_binding": True,
                "claims": [{"id": "birth_date", "path": ["birth_date"]}],
                "x_swiyu_zkp": {
                    "profile": PROFILE,
                    "circuit_id": CIRCUIT_ID,
                    "now_date": NOW_DATE,
                    "issuer_pub_x": issuer_x,
                    "issuer_pub_y": issuer_y,
                },
            }]
        },
        "proof_path": "/proof_b64",
        "variants": [
            _variant(VARIANT_A.variant_id, VARIANT_A),
            _variant(VARIANT_B.variant_id, VARIANT_B),
        ],
    }


def main() -> None:
    sys.stdout.write(json.dumps(build_document(), separators=(",", ":")) + "\n")


if __name__ == "__main__":
    main()

File: integration/providers/openac/transcript-fixture.mjs
#!/usr/bin/env node
/**
 * OpenAC transcript campaign factory.
 * Emits >=2 variants with the same issuer metadata and different hidden
 * birthdate + statusIndex. Does not claim semantic equivalence.
 */
import { createHash } from "node:crypto";
import { PROFILE } from "./paths.mjs";
import {
  CHECKED_IN_DID_TDW_KID,
  buildCredentialFixture,
  makeChallenge,
  makeStatus,
} from "./tests/fixture.mjs";

const CURRENT_TIME = String(Math.floor(Date.now() / 1000));
const CUTOFF = "2007-06-15";

function givenFor(fixture, status) {
  const challenge = { ...makeChallenge(status.snapshot.id), currentTime: CURRENT_TIME };
  return {
    issuer: {
      public_key: fixture.issuerPublicKey,
      issuer_id: "did:example:issuer",
      key_id: CHECKED_IN_DID_TDW_KID,
      allowed_vcts: ["https://example.ch/vct/person"],
    },
    current_time: CURRENT_TIME,
    age18_cutoff: CUTOFF,
    session: {
      nonce: challenge.nonce,
      client_id: challenge.clientId,
      response_uri: challenge.responseUri,
      state: challenge.state,
      query_id: challenge.queryId,
    },
    status_snapshot: {
      id: status.authoritativeSnapshot.id,
      issuer_id: status.authoritativeSnapshot.issuer,
      key_id: status.authoritativeSnapshot.kid,
      subject: status.authoritativeSnapshot.subject,
      commitment: BigInt(status.authoritativeSnapshot.commitment.hashHi).toString(16).padStart(32, "0") + BigInt(status.authoritativeSnapshot.commitment.hashLo).toString(16).padStart(32, "0"),
      epoch: String(status.authoritativeSnapshot.epoch),
      list_length: status.authoritativeSnapshot.listLength,
      valid_before: status.authoritativeSnapshot.validBefore,
    },
  };
}

function variant(id, options) {
  const fixture = buildCredentialFixture({
    swiyuIssuerShape: true,
    birthdate: options.birthdate,
    statusIndex: options.statusIndex,
  });
  const status = makeStatus(fixture.statusIndex);
  const challenge = { ...makeChallenge(status.snapshot.id), currentTime: CURRENT_TIME };
  const given = givenFor(fixture, status);
  return {
    id,
    credential: { format: "dc+sd-jwt", data: fixture.compactSdJwt },
    given,
    prepare: {
      profile: PROFILE,
      credential: { format: "dc+sd-jwt", data: fixture.compactSdJwt },
      context: { issuerPublicKey: fixture.issuerPublicKey },
    },
    present: {
      profile: PROFILE,
      request_context: challenge,
      inputs: {
        holderPrivateKeyHex: fixture.holderPrivateKeyHex,
        statusSnapshot: status.snapshot,
        statusWitness: status.witness,
      },
    },
    verify: {
      profile: PROFILE,
      request_context: challenge,
      inputs: {
        issuerPublicKey: fixture.issuerPublicKey,
        statusSnapshot: status.authoritativeSnapshot,
      },
    },
    compact_sd_jwt_sha256: createHash("sha256").update(fixture.compactSdJwt).digest("hex"),
  };
}

const document = {
  schema: "swiyu.transcript-fixture.v1",
  equivalence: "not_claimed",
  mapping: {
    format: "dc+sd-jwt",
    profile: PROFILE,
    query_id: "age-over-18-and-valid-status",
    challenge_from: "present.request_context",
    givens: {
      issuer: "issuer",
      current_time: "current_time",
      cutoff: "age18_cutoff",
      session: "session",
      status: "status_snapshot",
    },
    challenge_fields: {nonce: "nonce", clientId: "client_id", responseUri: "response_uri", state: "state"},
    signature_opaque: true,
  },
  dcql_query: { credentials: [{
    id: "age-over-18-and-valid-status", format: "dc+sd-jwt",
    meta: {vct_values: ["https://example.ch/vct/person"]},
    require_cryptographic_holder_binding: true,
    claims: [{id: "birthdate", path: ["birthdate"]}],
    x_swiyu_zkp: {profile: PROFILE, circuit_id: "swiyu_age18_status_2k", cutoff_date: CUTOFF, current_time: Number(CURRENT_TIME), status_list_snapshot: "ch-tsl-2026-07-epoch-172"}
  }]},
  proof_path: "/proof",
  variants: [
    variant("dob-2000-02-29-status-42", {
      birthdate: "2000-02-29",
      statusIndex: 42,
    }),
    variant("dob-1995-12-01-status-7", {
      birthdate: "1995-12-01",
      statusIndex: 7,
    }),
  ],
};

process.stdout.write(`${JSON.stringify(document)}\n`);
